File: 04_interface_segregation/typescript/index-bad.ts
interface Entity {
  attackDamage: number
  health: number
  name: string

  move (): any
  attack (targetEntity: Entity): any
  takeDamage (amount: number): any
}

class Character implements Entity {
  attackDamage: any
  health: any
  name: any
  move () {
    console.log(`${this.name} moved`)
  }
  attack (targetEntity: Entity) {
    console.log(
      `${this.name} attacked ${targetEntity.name} for ${this.attackDamage} damage`,
    )
    targetEntity.takeDamage(this.attackDamage)
  }
  takeDamage (amount: number) {
    this.health -= amount
    console.log(`${this.name} has ${this.health} health remaining`)
  }
}

class Turret implements Entity {
  attackDamage: any
  health: any
  name: any
  attack (targetEntity: Entity) {
    console.log(
      `${this.name} attacked ${targetEntity.name} for ${this.attackDamage} damage`,
    )
    targetEntity.takeDamage(this.attackDamage)
  }
  move () {
    throw new Error("Method not implemented.")
  }
  takeDamage (amount: number) {
    throw new Error("Method not implemented.")
  }
}

class Wall implements Entity {
  attackDamage: any
  health: any
  name: any
  move () {
    throw new Error("Method not implemented.")
  }
  attack () {
    throw new Error("Method not implemented.")
  }
  takeDamage (amount: number) {
    this.health -= amount
    console.log(`${this.name} has ${this.health} health remaining`)
  }
}

File: 04_interface_segregation/dart/index-bad.dart
abstract class Entity {
  late int attackDamage;
  late int health;
  late String name;

  void move();
  void attack(Entity targetEntity);
  void takeDamage(int amount);
}

class Character implements Entity {
  Character(this.name, this.attackDamage, this.health);

  int attackDamage;
  int health;
  String name;

  move() {
    print('${this.name} moved');
  }

  attack(Entity targetEntity) {
    print(
      '${this.name} attacked ${targetEntity.name} for ${this.attackDamage} damage',
    );
    targetEntity.takeDamage(this.attackDamage);
  }

  takeDamage(int amount) {
    this.health -= amount;
    print('${this.name} has ${this.health} health remaining');
  }
}

class Turret implements Entity {
  Turret(this.name, this.attackDamage);

  String name;
  int attackDamage;
  late int health;

  attack(Entity targetEntity) {
    print(
      '${this.name} attacked ${targetEntity.name} for ${this.attackDamage} damage',
    );
    targetEntity.takeDamage(this.attackDamage);
  }

  move() {
    throw Exception("Method not implemented.");
  }

  takeDamage(int amount) {
    throw Exception("Method not implemented.");
  }
}

class Wall implements Entity {
  Wall(this.name, this.health);

  late int attackDamage;
  int health;
  String name;

  move() {
    throw Exception("Method not implemented.");
  }

  attack(Entity targetEntity) {
    throw Exception("Method not implemented.");
  }

  takeDamage(int amount) {
    this.health -= amount;
    print('${this.name} has ${this.health} health remaining');
  }
}

void main(List<String> args) {
  var turret = Turret('Turret', 5);
  var character = Character('Character', 3, 100);
  var wall = Wall('Wall', 200);

  turret.attack(character);
  character.move();
  character.attack(wall);
  turret.attack(character);
}
